### ISSUE TITLE
Error allocating the fake swapchain to amdgpu device

### ISSUE THREAD

**Issue by WouterServaes** [ralgar/vkShade#19]
Running Red Dead Redemption 2.
With vkshade log level on debug, vkshade preset file and config file defined, and vkshade enabled.

Using this shader, renamed the ini to ReShade.ini, dropped the shaders from it in the correct folder.
https://www.nexusmods.com/reddeadredemption2/mods/4?tab=files&file_id=530 

No Wine error window, this is Steam's log.

No issues running the game without launch options.

log and context:
``` 
vkShade debug: activating mutable_format
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
Fossilize INFO: Overriding serialization path: "/PATH_TO_STEAM/shadercache/1174180/fozpipelinesv6/steamapprun_pipeline_cache".
vkShade debug: Found graphics capable queue
vkShade debug: 50 44
vkShade debug: format 44
vkShade debug: fake image size: 33587200
vkShade debug: fake image alignment: 65536
radv/amdgpu: Failed to allocate a buffer:
radv/amdgpu:    size      : 20171980800 bytes
radv/amdgpu:    alignment : 262144 bytes
radv/amdgpu:    domains   : 4
vkShade err:   ASSERT_VULKAN failed in ../src/vulkan/fake_swapchain.cpp : 74; -2
wine client error:54c: write: Bad file descriptor
pid 45981 != 45980, skipping destruction (fork without exec?)
Game 1174180 created interface STEAMAPPS_INTERFACE_VERSION008 / Apps
Game 1174180 created interface STEAMUSERSTATS_INTERFACE_VERSION011 / UserStats
Game 1174180 created interface SteamController004 / Controller
Game 1174180 created interface SteamFriends015 / Friends
Game 1174180 created interface SteamUser019 / User
Game 1174180 created interface SteamUtils008 / Utils
Game 1174180 method call count for IClientAppManager::GetCurrentLanguage : 5
Game 1174180 method call count for IClientUtils::RecordSteamInterfaceCreation : 8
Game 1174180 method call count for IClientUtils::GetAppID : 9
Game 1174180 method call count for IClientFriends::GetPersonaName : 1
Game 1174180 method call count for IClientUser::GetAuthSessionTicket : 2
Game 1174180 method call count for IClientUser::GetSteamID : 2
Game 1174180 method call count for IClientUser::BLoggedOn : 1
Uploaded AppInterfaceStats to Steam

(process:46444): GLib-GObject-CRITICAL **: 00:23:43.435: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
[2023-10-11 00:23:53] Background update loop checking for update. . . 
```

Happy to provide more info.

**Comment by WouterServaes** [comment]
Notice the amdgpu failing to allocate the buffer, causing the assertion. 
I'm running a AMD Radeon RX 6750 XT

**Comment by WouterServaes** [comment]
memoryRequirements.size becomes 33652736 at line 63.

No idea how any of this works, if anyone knows what the issue is I'd like to know. Goodluck.

**Comment by ralgar** [comment]
Oh boy, this is a good one! Thanks for the detailed info.

## The Issues

### Memory Allocation for the Swapchain Image

As you mentioned, it first appears that there's a discrepancy between what the layer determines the required alignment to be (64KB), versus what the RADV driver actually expects (256KB). However that is far from the only problem.

The actual Vulkan return code when trying to allocate the image is `VK_ERROR_OUT_OF_DEVICE_MEMORY`, indicating a lack of available GPU memory - which makes sense as it seems to be attempting to allocate almost 19GB (wtf?) for the swapchain image.

I'm not sure exactly where things are going wrong here at the moment, but I suspect this issue will be resolved as I refactor the existing Vulkan implementation (currently a slow moving WIP), and implement high quality abstractions like AMD's Vulkan Memory Allocator library.

### Invalid SPIR-V Code

I tested this configuration on my Nvidia GPU, and didn't run into the swapchain issue, so that would appear to be specific to either AMD GPUs or the RADV driver. I do get an error with the Lightroom shader, but the layer attempts to continue past that. I disabled that shader for good measure, but it doesn't appear to make a difference either way.

I then run into another crash, where the validation layers claim that one of the shaders doesn't compile to valid SPIR-V code (it has no entry point). It then appears to segfault when trying to access the descriptor binding that it didn't actually allocate for that invalid shader.

Again, I'm not entirely sure what is happening here at the moment, and the layer really isn't providing adequate debugging information at the moment. No indication of *which* shader is causing the problem.

---

```
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
vkShade debug: Found graphics capable queue
vkShade debug: 50 44
vkShade debug: format 44
vkShade debug: fake image size: 3866624
vkShade debug: fake image alignment: 1024
vkShade debug: created fake swapchain images
vkShade debug: current effectString Lightroom
vkShade debug: 3 images in firstImages
vkShade debug: not using swapchain images as second images
vkShade debug: 3 images in secondImages
vkShade debug: in creating ReshadeEffect
vkShade debug: created input ImageViews
vkShade debug: created ImageViews
vkShade err:   /home/ralgar/.config/vkShade/reshade-shaders/Shaders/qUINT_lightroom.fx(851, 1): error X4568: 'PS_ProcessLUT': input parameter 'huefactors' interpolation qualifiers do not match vertex shader ones
vkShade err:   /home/ralgar/.config/vkShade/reshade-shaders/Shaders/qUINT_lightroom.fx(857, 1): error X4568: 'PS_ApplyLUT': input parameter 'huefactors' interpolation qualifiers do not match vertex shader ones
vkShade err:   /home/ralgar/.config/vkShade/reshade-shaders/Shaders/qUINT_lightroom.fx(871, 1): error X4568: 'PS_DisplayStatistics': input parameter 'huefactors' interpolation qualifiers do not match vertex shader ones
vkShade debug: created reshade shaderModule
vkShade debug: frametime
vkShade debug: size: 4
vkShade debug: offset: 0
vkShade debug: Stencil Format: 129
vkShade debug: created descriptorSetLayouts
vkShade debug: created descriptorPool
vkShade debug: created Pipeline layout
vkShade debug: output writes: 0
vkShade debug: before writing buffer descriptor Sets
vkShade debug: before allocating descriptor Sets
vkShade debug: before writing descriptor Sets
vkShade debug: before writing descriptor Sets
vkShade debug: before writing descriptor Sets
vkShade debug: after writing ImageSamplerDescriptorSets
vkShade debug: finished creating Reshade effect
vkShade debug: created ReshadeEffect
vkShade debug: current effectString RTGlobalIllumination
vkShade debug: 3 images in firstImages
vkShade debug: not using swapchain images as second images
vkShade debug: 3 images in secondImages
vkShade debug: in creating ReshadeEffect
vkShade debug: created input ImageViews
vkShade debug: created ImageViews
VUID-VkShaderModuleCreateInfo-pCode-01379(ERROR / SPEC): msgNum: 706474367 - Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-01379 ] | MessageID = 0x2a1bf17f | SPIR-V module not valid: No OpEntryPoint instruction was found. This is only allowed if the Linkage capability is being used. The Vulkan spec states: If pCode is a pointer to GLSL code, it must be valid GLSL code written to the GL_KHR_vulkan_glsl GLSL extension specification (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-01379)
    Objects: 0
vkShade debug: created reshade shaderModule
vkShade debug: Stencil Format: 129
vkShade debug: created descriptorSetLayouts
VUID-VkDescriptorPoolSize-descriptorCount-00302(ERROR / SPEC): msgNum: 1303077118 - Validation Error: [ VUID-VkDescriptorPoolSize-descriptorCount-00302 ] | MessageID = 0x4dab60fe | vkCreateDescriptorPool(): pCreateInfo->pPoolSizes[0].descriptorCount is not greater than 0. The Vulkan spec states: descriptorCount must be greater than 0 (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkDescriptorPoolSize-descriptorCount-00302)
    Objects: 0
vkShade debug: created descriptorPool
vkShade debug: created Pipeline layout
vkShade debug: output writes: 0
zsh: segmentation fault (core dumped)  ENABLE_VKSHADE=1 VKSHADE_LOG_LEVEL=debug VK_INSTANCE_LAYERS= vkcube
```

---

If you like, you could try running the application again with the env var `VK_INSTANCE_LAYERS=VK_LAYER_KHRONOS_validation`. That *may* provide some additional context around the memory issue.

Also, please keep in mind this project is still in its very early stages. Most of the code base was inherited, and is badly in need of refactoring/updating, so this layer probably won't be in a stable state for some time yet. I do intend to bring it up to par though, and I appreciate the valuable bug report.

**Comment by WouterServaes** [comment]
Hello, thank you for the thorough reply! 
I think the shaders come from https://github.com/martymcmodding/qUINT, but no idea honestly. This was linked at the bottom of the nexusmods page.

Here is the entire log w/ VK_INSTANCE_LAYERS=VK_LAYER_KHRONOS_validation,
starting from application launch. I did not show this in the previous log, but the duplicated vkShade info happens every time.

Thank you for your work! 

```

/bin/sh\0-c\0VK_INSTANCE_LAYERS=VK_LAYER_KHRONOS_validation VKSHADE_PRESET_FILE=PATH_TO_STEAM/common/Red\ Dead\ Redemption\ 2/ReShade.ini VKSHADE_CONFIG_FILE=~/.config/vkShade/vkShade.conf VKSHADE_LOG_LEVEL=debug  ENABLE_VKSHADE=1 HOME_DIR/.local/share/Steam/ubuntu12_32/reaper SteamLaunch AppId=1174180 -- HOME_DIR/.local/share/Steam/ubuntu12_32/steam-launch-wrapper -- 'HOME_DIR/.local/share/Steam/steamapps/common/SteamLinuxRuntime_sniper'/_v2-entry-point --verb=waitforexitandrun -- 'HOME_DIR/.local/share/Steam/steamapps/common/Proton - Experimental'/proton waitforexitandrun  '/PATH_TO_STEAM/common/Red Dead Redemption 2/PlayRDR2.exe'\0
chdir "/PATH_TO_STEAM/common/Red Dead Redemption 2"
ERROR: ld.so: object 'HOME_DIR/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object 'HOME_DIR/.local/share/Steam/ubuntu12_64/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS64): ignored.
ERROR: ld.so: object 'HOME_DIR/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object 'HOME_DIR/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object 'HOME_DIR/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
fsync: up and running.
wine: RLIMIT_NICE is <= 20, unable to use setpriority safely
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
Setting breakpad minidump AppID = 1174180
Steam_SetMinidumpSteamID:  Caching Steam ID:  76561198072946103 [API loaded no]
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
Missing interface adapter for SteamController008 UGC
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
Fossilize INFO: Overriding serialization path: "/PATH_TO_STEAM/shadercache/1174180/fozpipelinesv6/steamapprun_pipeline_cache".
vkShade debug: Found graphics capable queue
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
Fossilize INFO: Overriding serialization path: "/PATH_TO_STEAM/shadercache/1174180/fozpipelinesv6/steamapprun_pipeline_cache".
vkShade debug: Found graphics capable queue
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
wine: Read access denied for device L"\\??\\D:\\", FS volume label and serial are not available.
wine: Read access denied for device L"\\??\\E:\\", FS volume label and serial are not available.
wine: Read access denied for device L"\\??\\F:\\", FS volume label and serial are not available.
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
10/11 22:39:03 Init: Installing breakpad exception handler for appid(gameoverlayui)/version(20230929173826)/tid(24853)
10/11 22:39:03 Init: Installing breakpad exception handler for appid(gameoverlayui)/version(1.0)/tid(24853)
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
Fossilize INFO: Overriding serialization path: "/PATH_TO_STEAM/shadercache/1174180/fozpipelinesv6/steamapprun_pipeline_cache".
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
Fossilize INFO: Overriding serialization path: "/PATH_TO_STEAM/shadercache/1174180/fozpipelinesv6/steamapprun_pipeline_cache".
vkShade debug: Found graphics capable queue
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
Fossilize INFO: Overriding serialization path: "/PATH_TO_STEAM/shadercache/1174180/fozpipelinesv6/steamapprun_pipeline_cache".
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade debug: DestroyCommandPool
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
vkShade debug: Found graphics capable queue
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade debug: DestroyCommandPool
vkShade info:  config file: HOME_DIR/.config/vkShade/vkShade.conf
vkShade info:  reshadeTexturePath = /opt/reshade-shaders/Textures
vkShade info:  reshadeIncludePath = /opt/reshade-shaders/Shaders
vkShade info:  toggleKey = Home
vkShade info:  enableOnLaunch = True
vkShade info:  depthCapture = off
vkShade info:  Preset file: PATH_TO_STEAM/common/Red Dead Redemption 2/ReShade.ini
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
Fossilize INFO: Overriding serialization path: "/PATH_TO_STEAM/shadercache/1174180/fozpipelinesv6/steamapprun_pipeline_cache".
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
vkShade debug: Found graphics capable queue
vkShade debug: DestroyCommandPool
vkShade debug: device supports VK_KHR_swapchain_mutable_format
vkShade debug: activating mutable_format
Fossilize INFO: Overriding serialization path: "/PATH_TO_STEAM/shadercache/1174180/fozpipelinesv6/steamapprun_pipeline_cache".
vkShade debug: Found graphics capable queue
vkShade debug: 50 44
vkShade debug: format 44
vkShade debug: fake image size: 33587200
vkShade debug: fake image alignment: 65536

radv/amdgpu: Failed to allocate a buffer:
radv/amdgpu:    size      : 20171980800 bytes
radv/amdgpu:    alignment : 262144 bytes
radv/amdgpu:    domains   : 4
vkShade err:   ASSERT_VULKAN failed in ../src/vulkan/fake_swapchain.cpp : 75; -2
wine client error:548: write: Bad file descriptor

(process:24853): GLib-GObject-CRITICAL **: 22:39:08.813: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
pid 24314 != 24313, skipping destruction (fork without exec?)
Game 1174180 created interface STEAMAPPS_INTERFACE_VERSION008 / Apps

```

**Comment by ralgar** [comment]
After a second look, it appears I was misinterpreting the number of shaders intended to be loaded with this preset. When I implemented the preset parsing, I simply used the `TechniquesSorting` key to determine which shaders to load, and in which order. I see now that this is not always a viable solution.

It looks like it is possible for ReShade presets to have many shaders in the `TechniquesSorting` key, but only a small handful actually enabled via the `Techniques` key. I'll have to update the layer to handle this situation, but it shouldn't be difficult.

>Here is the entire log w/ VK_INSTANCE_LAYERS=VK_LAYER_KHRONOS_validation,
starting from application launch. I did not show this in the previous log, but the duplicated vkShade info happens every time.

Thank you for this! What a mess with the duplication though. As far as I can tell that isn't an issue of the layer itself, but possibly a quirk of Proton initializing Vulkan multiple times, for shader compiling and possibly other pre-launch tasks. Just a guess, unfortunately I don't know as much about Proton as I'd like.

Anyway, it looks like there are no validation errors at least.

If it's alright with you, I'll ping you after I refactor the Vulkan implementation, and have you test this memory allocation bug again. Unfortunately I don't have an AMD GPU available, so I won't be able to reproduce it myself.
